File: elearning1/code/go/tcp_tuning.go
package main

import (
	"bufio"
	"fmt"
	"net"
	"os"
	"time"
)

func main() {
	if len(os.Args) < 4 {
		fmt.Println("Usage: go run tcp_tuning.go <host> <port> <message>")
		return
	}
	host := os.Args[1]
	port := os.Args[2]
	message := os.Args[3]

	addr, err := net.ResolveTCPAddr("tcp", net.JoinHostPort(host, port))
	if err != nil {
		panic(err)
	}

	conn, err := net.DialTCP("tcp", nil, addr)
	if err != nil {
		panic(err)
	}
	defer conn.Close()

	conn.SetNoDelay(true)
	conn.SetKeepAlive(true)
	conn.SetKeepAlivePeriod(30 * time.Second)
	conn.SetReadBuffer(1 << 16)
	conn.SetWriteBuffer(1 << 16)

	start := time.Now()
	if _, err := conn.Write([]byte(message)); err != nil {
		panic(err)
	}
	conn.SetReadDeadline(time.Now().Add(10 * time.Second))
	r := bufio.NewReader(conn)
	resp, err := r.ReadString('\n')
	if err != nil {
		// Try reading bytes if newline wasn't sent
		b := make([]byte, 1024)
		n, _ := conn.Read(b)
		resp = string(b[:n])
	}
	fmt.Printf("Echo: %s\n", resp)
	fmt.Printf("Approx RTT: %.2f ms\n", time.Since(start).Seconds()*1000)
}

File: elearning1/code/nodejs/tcp_tuning.js
const net = require('net');

if (process.argv.length < 5) {
  console.log('Usage: node tcp_tuning.js <host> <port> <message>');
  process.exit(0);
}

const host = process.argv[2];
const port = parseInt(process.argv[3], 10);
const message = process.argv[4];

const socket = net.connect({ host, port }, () => {
  socket.setNoDelay(true);
  socket.setKeepAlive(true, 30000);
  socket.setTimeout(10000);
  const start = Date.now();
  socket._startTime = start;
  socket.write(message + '\n');
});

socket.on('data', (data) => {
  const rtt = Date.now() - (socket._startTime || Date.now());
  console.log('Echo:', data.toString());
  console.log(`Approx RTT: ${rtt.toFixed(2)} ms`);
  socket.end();
});

socket.on('timeout', () => {
  console.error('Read timeout');
  socket.destroy();
});

socket.on('error', (err) => {
  console.error('Socket error:', err.message);
});
